File: tests/solutions_cpp/top_k_frequent_elements.cpp
#include "json_helper.h"

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> count;
        for (int n : nums) count[n]++;
        // Min-heap of size k by frequency
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        for (auto& [val, cnt] : count) {
            pq.push({cnt, val});
            if ((int)pq.size() > k) pq.pop();
        }
        vector<int> result;
        while (!pq.empty()) {
            result.push_back(pq.top().second);
            pq.pop();
        }
        sort(result.begin(), result.end());
        return result;
    }
};

int main() {
    auto args = parse_args();
    Solution sol;
    auto nums = to_vector_int(args[0]);
    int k = to_int(args[1]);
    auto result = sol.topKFrequent(nums, k);
    cout << to_json(result) << endl;
    return 0;
}

File: tests/solutions_js/top_k_frequent_elements.js
class Solution {
  topKFrequent(nums, k) {
    const count = new Map();
    for (const num of nums) {
      count.set(num, (count.get(num) || 0) + 1);
    }

    // Bucket sort: index = frequency, value = list of nums with that frequency
    const buckets = Array.from({ length: nums.length + 1 }, () => []);
    for (const [num, freq] of count) {
      buckets[freq].push(num);
    }

    const result = [];
    for (let i = buckets.length - 1; i >= 0 && result.length < k; i--) {
      for (const num of buckets[i]) {
        result.push(num);
        if (result.length === k) break;
      }
    }

    return result;
  }
}
